### ISSUE TITLE
Performance

### ISSUE THREAD

**Issue by sagikazarmark** [casbin/casbin#45]
I'm still trying to get my head around the model of casbin, but I already have some concerns about it's performance when used with a lot of users.

Given I have an increasing number of users (1000+) and an increasing number of entities (1000+) and I want to grant access to these entities based on certain rules (casbin), how can I ensure that my Go application won't behave as a Java application at some point eating up all the available RAM in the whole data center?

As far as I can see casbin always loads the whole set of rules into the memory, there is no dynamic lookup.

Another concern regarding the number of entities/users: the more data I have, the more changes I receive meaning I have to reload policies frequently.

How does casbin perform in such scenarios? Can this mean that casbin might not be a perfect fit for my use case?

Thanks in advance!

**Comment by hsluoyz** [comment]
This is also a question that concerns me and asked many times:) How many rules will be there?  If your scenario don't allow users themselves to edit the rules, there shouldn't be too many rules.

**Comment by sagikazarmark** [comment]
> If your scenario don't allow users themselves to edit the rules, there shouldn't be too many rules.

I don't see how this is true, but let me give an example:

Users belong to multiple groups, or I should say they are members of groups. There are two types of memberships: owners and regular members.

All members can create content in a group, but only owners and content creators can edit content.

Owners can manage user membership.

There is a limited number of owners within a group, but there is no limit on the number of groups and there is no limit on memberships.

So potentially there is an increasing number of users and groups (meaning memberships as the connection between them), within groups there is an increasing number of content to which we have to apply the rules above (group owners and content owners can edit content)

As I said, I still have to get my head around casbin's model, but I believe an ABAC based authorization would better fit my scenario as I can clearly describe connections between objects and can define rules based on them instead of some list/role based authorization.

What do you think?

The ACL/RBAC performance still concerns me (although I probably don't need them for the moment, I would hate building an entirely separate logic just for RBAC while I have a great library pulled in for authorization). Are there any examples/benchmarks for large scale use (eg. the scenario I outlined above with RBAC)?

**Comment by hsluoyz** [comment]
I have benchmarked the overhead of ``Enforce()`` call here: https://github.com/casbin/casbin#benchmarks. Do you think such a delay is acceptable? I will add memory usage data later here.

**Comment by ntdunglc** [comment]
For RABC (large), I think it can be considered slow since it may be slower than a network call.
In enforcer.go, I dont see the expression of govaluate is cached which is against goevaluate's documentation, is there any reason it's needed to work this way? 
I haven't profile the code so not sure if it's the bottleneck.

**Comment by hsluoyz** [comment]
Casbin doesn't do the expression caching at its side. I believe govaluate doesn't do either. I have asked the author about advice to upgrade performance.

**Comment by Knetic** [comment]
I spent some time running through the benchmarks for casbin, and after a few runs with CPU profiling engaged, it seems as if the biggest bottleneck is `Model.HasPolicy`, with 82% of the benchmark's time spent executing that function. `(*Enforcer).Enforce` (where govaluate is used) takes about 13% of the time, which is probably not significant enough for the optimizations around LRU/type checking to show large overall gains (if it only takes 13%, and LRU/type checking reduce the time by 10%, that's only a 1.3% overall improvement, which is hardly noticeable even at these scales).

Attached are a couple of the CPU profiles ([benchmark 1](https://user-images.githubusercontent.com/5728164/32740176-ef2c91a4-c856-11e7-860d-cfcbc057d434.png) and [benchmark 2](https://user-images.githubusercontent.com/5728164/32740177-ef3d52fa-c856-11e7-9116-995b6d631828.png)) obtained with the following;

```
go test -bench=BenchmarkRBACModel -cpuprofile=/tmp/cpu_untyped1.out -run=adrhwerhafb
go tool pprof -svg /tmp/casbin_cpu1.out > /tmp/casbin_cpu1.svg
```

The funny `-run` is to just skip the unit tests, since i only cared about benchmarks. I converted the to PNG's manually because github apparently doesn't support SVG images in comments.

**Comment by hsluoyz** [comment]
Thanks for looking at this issue:) I saw the CPU profiles. ``Model.HasPolicy`` indeed takes up a lot of time. And it's called by ``AddPolicy()`` and ``AddGroupingPolicy()`` in the test. These calls belong to the initialization stage. But I have called ``b.ResetTimer()`` to reset the benchmark timer. So when ``(*Enforcer).Enforce`` is benchmarked, the timer starts from zero. Based on Golang's docs, I think ``b.ResetTimer()`` is just used for initializing some data before test.

https://github.com/casbin/casbin/blob/13b7cf40bc982c9dd1bf2c026ee6bd220267baee/model_b_test.go#L96-L114

**Comment by MaerF0x0** [comment]
Imagine trying to apply this library to something like Google's drive: 

millions of users grant view, edit, comment access to documents+user combos or publicly  across billions entities.

Could casbin with a DB persistence work for this?

**Comment by hsluoyz** [comment]
@MaerF0x0 , Casbin can do this with appropriate sharding (e.g. sharding based on region).

**Comment by jon-whit** [comment]
@hsluoyz - Assuming each request is safeguarded with fine-grained authorization using Casbin, and no sharding is applied, do you anticipate that Casbin could handle roughly 20 million requests per day without a severe bottleneck?

**Comment by hsluoyz** [comment]
@jon-whit Only one Casbin instance cannnot handle such many requests. But Casbin supports the coordination between multiple Casbin enforcer nodes via [etcd-watcher](https://github.com/casbin/etcd-watcher). Overall, you will need:

1. Multiple Casbin enforcer instances (each instance runs on a machine or virtual machine)
2. A coordination mechanism like etcd (use the official etcd-watcher or ZooKeeper which you can implement it as a watcher). This is used to synchronize the policy rules.
3. Backend storage for policy rules such as Redis, Memcached, Chef or even cloud storage like Amazon S3, Azure Blob Storage, etc. The core idea is to make the storage scalable too. Casbin already has an official Redis adapter, for other types of storage I mentioned, you may need to implement an adapter, which is a quite easy job. For a large number of requests requirement as yours, I think you already have a distributed backend storage. So just use that to store Casbin policy rules too.

**Comment by murrays** [comment]
I replicated the results using casbin.Net.  The performance appears to degrade linearly depending on the size of the rules table. 
Looking at the implementation this is because the policy objects are essentially List<List<string>> so there is no opportunity to prune the number of records hit.  I did wonder if the Policy object could be replaced with HashMap<string> (or Dictionary<string,List<string>> in C# parlance).  However the 'function' evaluation allows for keying on any column.  

I tried the 'MED' scale test for RBAC with 10000 rules and got 43ms per operation.  That's not fast.  I didn't launch the 100000 version of the test as it would have taken quite some time!    I'm looking at potential rules table sizes of around 1m and projecting these results I would be looking at 300ms per operation!  

As a comparison, .Net Core Identity has a flat response time of around 1.3ms/op tested at 100/1000/10000/100000 rules.  Maybe its worth looking at a cross platform library to use the Identity schema?

Note that both my sets of tests used a PostgreSQL back end.  However whatever the implementation, the underlying 'schema' allows for indexing which I believe leads to the better performance profile with larger datasets.
